src/analysis/context.ml

```diff
@@ -46,15 +48,28 @@ type t =
 
 let to_string = function
   | Constructor (cd, _) -> Printf.sprintf "constructor %s" cd.cstr_name
+  | Unknown_constructor -> Printf.sprintf "unknown constructor"
   | Expr -> "expression"
   | Label lbl -> Printf.sprintf "record field %s" lbl.lbl_name
+  | Unknown_label -> Printf.sprintf "record field"
   | Module_path -> "module path"
   | Module_type -> "module type"
   | Patt -> "pattern"
   | Constant -> "constant"
   | Type -> "type"
   | Unknown -> "unknown"
 
+let of_locate_context : Query_protocol.Locate_context.t -> t = function
+  | Expr -> Expr
+  | Module_path -> Module_path
+  | Module_type -> Module_type
+  | Patt -> Patt
+  | Type -> Type
+  | Constant -> Constant
+  | Constructor -> Unknown_constructor
+  | Label -> Unknown_label
+  | Unknown -> Unknown
+
```

Comment:
Is it really useful to introduce the new `Query_protocol.Locate_context.t` ? 
I think you should simply rely on `Context.t`:
- The query parameter would stay a string option.
- You add a `Context.of_string` that does the parsing of such a user-chosen context.
- Either in `Query_command` or `Locate` you use that function to parse the string and continue with the `Context.t` that the user chose.

Reply:
I'm hesitant to use a `string` to represent the context in the `query_protocol` because I like the type-safety of `Locate_context.t`. And I considered using `Context.t` in `query_protocol`, but I decided against it for two reasons:
1. It would complicate the dependency structure
2. `Context.t` can contain a `Types.constructor_description` or `Types.label_description`. Since we want to be able to serialize queries (in `query_jsonl.ml`), this would complicate things.

Reply:
Is it really more type-safe though ? 🙂

I completely understand the appeal of parsing the user input as soon as possible and then only manipulate more descriptive types, but the duplication of the type with slight variations looks redundant and a bit confusing. If sharing the type is not easily doable then I would stick with the simpler version that (as multiple other commands do) simply forwards the `string` for later parsing.

Feel free to ignore this of course, it's really a matter of opinion.

(in both cases, you can use an alternative `to_string` function for `query_json`)

Reply:
Correct me if I'm wrong, but I believe the LSP communicates with Merlin via the `Query_protocol`? If that's the case, then there is a type-safety difference for that case.

Reply:
It simply changes the place were we deal with the unsafe user input: you push that task to the user of the API (lsp for example) while I suggest having that parsing done internally by the command. This has no relation to being type-safe or not, it's a matter of API design. Again, as I said earlier, I don't have a strong preference, so let's keep it that way if you prefer.